src/graph/graph_apis.cc

```diff
@@ -33,13 +35,25 @@ PackedFunc ConvertEdgeArrayToPackedFunc(const Graph::EdgeArray& ea) {
   return PackedFunc(body);
 }
 
+// Convert CSRArray structure to PackedFunc.
+PackedFunc ConvertAdjToPackedFunc(const std::vector<IdArray>& ea) {
+  auto body = [ea] (DGLArgs args, DGLRetValue* rv) {
+      const int which = args[0];
+      if ((size_t) which < ea.size()) {
+        *rv = std::move(ea[which]);
+      } else {
+        LOG(FATAL) << "invalid choice";
+      }
+    };
+  return PackedFunc(body);
+}
+
 // Convert Subgraph structure to PackedFunc.
 PackedFunc ConvertSubgraphToPackedFunc(const Subgraph& sg) {
   auto body = [sg] (DGLArgs args, DGLRetValue* rv) {
       const int which = args[0];
       if (which == 0) {
-        Graph* gptr = new Graph();
-        *gptr = std::move(sg.graph);
+        GraphInterface* gptr = sg.graph->Reset();
```

Comment:
I agree that the current solution is ideal. But I don't know what is a better solution.

```C++
GraphInterface* MoveGraph(GraphPtr gptr);
```
How to implement this? we need to explicitly cast the pointer to one of the graph type (Graph or ImmutableGraph). I don't feel comfortable with this solution either.

Reply:
Wait. Isn't GraphPtr alias of `std::shared_ptr<GraphInterface>`? Then move it to GraphInterface* is simply what's your code is doing right?

Reply:
gptr still owns the object. if gptr disappears, the memory is gone. We don't have any place to store gptr.

Reply:
See your code:
```c++
virtual GraphInterface *Reset() {
  Graph* gptr = new Graph();
  *gptr = std::move(*this);
  return gptr;
}
# in c_api.cc
GraphInterface* gptr = sg.graph->Reset();  # this sg.graph is a GraphPtr type
```
After Reset, the memory pointed by the shared pointer has already gone. What I proposed is exactly the same as this logic, but just refactor it to a global function.

Reply:
A proposed implementation of this function:
```c++
GraphInterface* MoveGraph(GraphPtr gptr) {
  Graph* ret = new Graph();
  *ret = std::move(*gptr);
  return ret;
}

Reply:
I see the problem of inheritance here and why you do it in a member function. As I've mentioned in another comment, since python only accepts raw pointers, so in order to fix this, is to return python a **pointer to the shared pointer**:
```
GraphPtr* ret = new GraphPtr;
*ret = sg.graph;
```
**Good thing:** Now if any C++ side code holds the shared pointer, it could be preserved without being reset.
**Bad thing:** Then all other graph apis need to be changed accordingly. Also the pointer to the pointer is kind of confusing.

That's why I directly use `Graph` object previously.

Reply:
The problem is that we can't use `Graph` object here.

Reply:
After discussing different options, I still don't see why my current solution is worse than others. There isn't a good solution here. `Reset` may not be a good name. Something like MoveGraph can be an option.